src/mono/mono/metadata/native-library.c

```diff
@@ -1223,11 +1223,32 @@ mono_loader_install_pinvoke_override (PInvokeOverrideFn override_fn)
 	pinvoke_override = override_fn;
 }
 
+static int mono_precompute_size (const char *key)
+{
+	int size = 1; // Null terminator
+	int len = (int)strlen (key);
+	for (int i = 0; i < len; ++i) {
+		unsigned char b = key[i];
+		if ((b >= '0' && b <= '9') ||
+		    (b >= 'a' && b <= 'z') ||
+			(b >= 'A' && b <= 'Z') ||
+			b == '_' || b == '.' || b == '-' || b ==  '+' || b == '<' || b == '>') {
+			size++;
+		}
+		else {
+			size += 4;
+		}
+	}
+	return size;
+}
+
 // Keep synced with FixupSymbolName from src/tasks/Common/Utils.cs
-char* mono_fixup_symbol_name (char *key) {
-	char* fixedName = malloc(256);
+char* mono_fixup_symbol_name (const char *prefix, const char *key, const char *suffix) {
+	int size = mono_precompute_size (key) + strlen (prefix) + strlen (suffix);
+	char fixedName[size];
```

Comment:
this creates a stack-allocated array of `size` characters.  then you return a pointer to this temporary buffer from this function.  that won't work.

You can heap allocate it.
```c
	char *fixedName = g_new0(char, size);
```

But a better idea is to use the `GString` type from `glib.h` and avoid using `strcpy`, `snprintf`, etc.

```c
   GString *str = g_string_sized_new (size);
   g_string_append_printf (str, "%s", prefix);
   
   for (int i = 0; i < len; ++i) {
       char b = key[i];
       if (is_symbol_char_verbatim (b)) {
            g_string_append_c (str, b);
       } else if (is_symbol_char_underscore (b)) {
            g_string_append_c (str, '_');
       } else {
            // Append the hex representation of b between underscores
            g_string_append_printf (str, "_%X_", b);
       }
   }
   g_string_append_printf (str, "%s", suffix);
   return g_string_free (str, FALSE);

Reply:
Thanks, updated!